File: sdkwork-rtc-sdk-csharp/src/SDKWork.Rtc.Sdk/RtcProviderPackageLoader.cs
namespace Sdkwork.Rtc.Sdk;

using System;
using System.Collections.Generic;
using System.Linq;

public sealed class RtcProviderPackageLoaderException : Exception
{
    public RtcProviderPackageLoaderException(string code, string message)
        : base(message)
    {
        this.code = code;
    }

    public string code { get; }
}

public sealed record RtcProviderPackageLoadRequest(
    string? providerKey = null,
    string? packageIdentity = null
);

public sealed record RtcResolvedProviderPackageLoadTarget(
    RtcProviderPackageCatalogEntry packageEntry
);

public delegate object? RtcProviderPackageImportFn(RtcResolvedProviderPackageLoadTarget target);
public delegate object? RtcProviderPackageLoaderFn(RtcProviderPackageLoadRequest request);

public sealed record RtcProviderPackageInstallRequest(
    object driverManager,
    RtcProviderPackageLoadRequest loadRequest
);

public static class RtcProviderPackageLoader
{
    public const string ProviderPackageNotFound = "provider_package_not_found";
    public const string ProviderPackageIdentityMismatch = "provider_package_identity_mismatch";
    public const string ProviderPackageLoadFailed = "provider_package_load_failed";
    public const string ProviderModuleExportMissing = "provider_module_export_missing";

    public static RtcResolvedProviderPackageLoadTarget ResolveRtcProviderPackageLoadTarget(
        RtcProviderPackageLoadRequest? request
    )
    {
        request ??= new RtcProviderPackageLoadRequest();
        var packageByProviderKey = string.IsNullOrWhiteSpace(request.providerKey)
            ? null
            : RtcProviderPackageCatalog.GetRtcProviderPackageByProviderKey(request.providerKey!);
        var packageByIdentity = string.IsNullOrWhiteSpace(request.packageIdentity)
            ? null
            : RtcProviderPackageCatalog.GetRtcProviderPackageByPackageIdentity(request.packageIdentity!);

        if (packageByProviderKey is not null
            && packageByIdentity is not null
            && !string.Equals(packageByProviderKey.packageIdentity, packageByIdentity.packageIdentity, StringComparison.Ordinal))
        {
            throw new RtcProviderPackageLoaderException(
                ProviderPackageIdentityMismatch,
                "providerKey and packageIdentity must resolve to the same provider package boundary."
            );
        }

        var resolvedPackage = packageByProviderKey ?? packageByIdentity;
        if (resolvedPackage is null)
        {
            throw new RtcProviderPackageLoaderException(
                ProviderPackageNotFound,
                "No official provider package matches the requested provider boundary."
            );
        }

        return new RtcResolvedProviderPackageLoadTarget(resolvedPackage);
    }

    public static RtcProviderPackageLoaderFn CreateRtcProviderPackageLoader(
        RtcProviderPackageImportFn importPackage
    ) => request => LoadRtcProviderModule(request, importPackage);

    public static object? LoadRtcProviderModule(
        RtcProviderPackageLoadRequest request,
        RtcProviderPackageImportFn importPackage
    )
    {
        var target = ResolveRtcProviderPackageLoadTarget(request);

        try
        {
            var providerModule = importPackage(target);
            if (providerModule is null)
            {
                throw new RtcProviderPackageLoaderException(
                    ProviderModuleExportMissing,
                    "Reserved provider package loader scaffold requires an executable provider module namespace."
                );
            }

            return providerModule;
        }
        catch (RtcProviderPackageLoaderException)
        {
            throw;
        }
        catch (Exception error)
        {
            throw new RtcProviderPackageLoaderException(
                ProviderPackageLoadFailed,
                $"Reserved provider package loader scaffold could not load {target.packageEntry.packageIdentity}: {error.Message}"
            );
        }
    }

    public static void InstallRtcProviderPackage(
        RtcProviderPackageInstallRequest request,
        RtcProviderPackageImportFn importPackage
    )
    {
        _ = LoadRtcProviderModule(request.loadRequest, importPackage);

        throw new RtcProviderPackageLoaderException(
            ProviderPackageLoadFailed,
            "Reserved provider package installer scaffold cannot register provider modules until a verified runtime bridge lands."
        );
    }

    public static void InstallRtcProviderPackages(
        IReadOnlyList<RtcProviderPackageInstallRequest> requests,
        RtcProviderPackageImportFn importPackage
    )
    {
        foreach (var request in requests)
        {
            _ = LoadRtcProviderModule(request.loadRequest, importPackage);
        }

        if (requests.Count > 0)
        {
            throw new RtcProviderPackageLoaderException(
                ProviderPackageLoadFailed,
                "Reserved provider package installer scaffold cannot register provider modules until a verified runtime bridge lands."
            );
        }
    }
}

File: sdkwork-rtc-sdk-flutter/lib/src/rtc_provider_package_loader.dart
import 'rtc_provider_package_catalog.dart';

final class RtcProviderPackageLoaderException implements Exception {
  const RtcProviderPackageLoaderException(this.code, this.message);

  final String code;
  final String message;

  @override
  String toString() => 'RtcProviderPackageLoaderException($code): $message';
}

final class RtcProviderPackageLoadRequest {
  const RtcProviderPackageLoadRequest({
    this.providerKey,
    this.packageIdentity,
  });

  final String? providerKey;
  final String? packageIdentity;
}

final class RtcResolvedProviderPackageLoadTarget {
  const RtcResolvedProviderPackageLoadTarget({
    required this.packageEntry,
  });

  final RtcProviderPackageCatalogEntry packageEntry;
}

typedef RtcProviderModuleNamespace = Object?;
typedef RtcProviderPackageImportFn = RtcProviderModuleNamespace Function(
  RtcResolvedProviderPackageLoadTarget target,
);
typedef RtcProviderPackageLoader = RtcProviderModuleNamespace Function(
  RtcProviderPackageLoadRequest request,
);

final class RtcProviderPackageInstallRequest {
  const RtcProviderPackageInstallRequest({
    required this.driverManager,
    required this.loadRequest,
  });

  final Object driverManager;
  final RtcProviderPackageLoadRequest loadRequest;
}

RtcResolvedProviderPackageLoadTarget resolveRtcProviderPackageLoadTarget(
  RtcProviderPackageLoadRequest request,
) {
  final packageByProviderKey = request.providerKey == null
      ? null
      : getRtcProviderPackageByProviderKey(request.providerKey!);
  final packageByIdentity = request.packageIdentity == null
      ? null
      : getRtcProviderPackageByPackageIdentity(request.packageIdentity!);

  if (packageByProviderKey != null &&
      packageByIdentity != null &&
      packageByProviderKey.packageIdentity != packageByIdentity.packageIdentity) {
    throw const RtcProviderPackageLoaderException(
      'provider_package_identity_mismatch',
      'providerKey and packageIdentity must resolve to the same provider package boundary.',
    );
  }

  final resolvedPackage = packageByProviderKey ?? packageByIdentity;
  if (resolvedPackage == null) {
    throw const RtcProviderPackageLoaderException(
      'provider_package_not_found',
      'No official provider package matches the requested provider boundary.',
    );
  }

  return RtcResolvedProviderPackageLoadTarget(packageEntry: resolvedPackage);
}

RtcProviderPackageLoader createRtcProviderPackageLoader({
  required RtcProviderPackageImportFn importPackage,
}) {
  return (request) => loadRtcProviderModule(
        request,
        importPackage: importPackage,
      );
}

RtcProviderModuleNamespace loadRtcProviderModule(
  RtcProviderPackageLoadRequest request, {
  required RtcProviderPackageImportFn importPackage,
}) {
  final target = resolveRtcProviderPackageLoadTarget(request);

  try {
    final namespace = importPackage(target);
    if (namespace == null) {
      throw const RtcProviderPackageLoaderException(
        'provider_module_export_missing',
        'Reserved provider package loader scaffold requires an executable provider module namespace.',
      );
    }

    return namespace;
  } on RtcProviderPackageLoaderException {
    rethrow;
  } catch (error) {
    throw RtcProviderPackageLoaderException(
      'provider_package_load_failed',
      'Reserved provider package loader scaffold could not load ${target.packageEntry.packageIdentity}: $error',
    );
  }
}

void installRtcProviderPackage(
  RtcProviderPackageInstallRequest request, {
  required RtcProviderPackageImportFn importPackage,
}) {
  loadRtcProviderModule(
    request.loadRequest,
    importPackage: importPackage,
  );

  throw const RtcProviderPackageLoaderException(
    'provider_package_load_failed',
    'Reserved provider package installer scaffold cannot register provider modules until a verified runtime bridge lands.',
  );
}

void installRtcProviderPackages(
  Iterable<RtcProviderPackageInstallRequest> requests, {
  required RtcProviderPackageImportFn importPackage,
}) {
  final materializedRequests = requests.toList(growable: false);
  for (final request in materializedRequests) {
    loadRtcProviderModule(
      request.loadRequest,
      importPackage: importPackage,
    );
  }

  if (materializedRequests.isNotEmpty) {
    throw const RtcProviderPackageLoaderException(
      'provider_package_load_failed',
      'Reserved provider package installer scaffold cannot register provider modules until a verified runtime bridge lands.',
    );
  }
}
